Add new actions API to the plugin

BastiaanOlij opened the pull request:
OpenVR introduced a new actions based API. 

> Read more about this in the documentation: https://github.com/GodotVR/godot-openvr-asset/wiki/OpenVR-actions (WIP)

Now that this is starting to work some basic info for now. I'll expand on this shortly.

The actions system in OpenVR is declarative and the moment you start using it, it disables support for the current way of working. For this reason the default manifest file provided in the project declares every button as is and the plugin reacts to these to mimic the old way of working with the only limitation that you can only do so for the left and right hand controllers.

**note** that tracking for additional controllers still works fine but buttons not. 
Also the default bindings files are incomplete at the moment so the first time you run your project OpenVR will ask you to setup your bindings (I was very impressed with the interface Valve came up with here)

The manifest and binding files can be found in: `res://addons/godot-openvr/actions/actions.json`

For many people especially those with existing VR projects or projects that target more than just the OpenVR platform this will be perfect. But those who want to use more will get stuck.

These developers can edit the manifest file and add there own actions. You can also remove the existing actions if you do not wish to use those. The plugin will simply disable support for the related feature.

You can change the supplied manifest file or you can tell the plugin where to look for yours by setting `OpenVRConfig.action_json_path`
You must do this BEFORE you call initialize!

Anytime **after** you have initialized you can register your actions and provide a callback that will be called when the action is trigger. (Note: I am still working on this part).

BastiaanOlij commented:
Did some major refactoring of the process function.

The controller handling is now completely internalized into openvr_data, made a new structure to track controller data.

Next up is binding default actions for those controllers so we get button feedback again.

BastiaanOlij commented:
Got a bunch further with loading the handles and stuff but currently it seems to hang after initialisation with a couple of transforms being messed up. I'm guessing its because I'm still using the old style of pose updates while I'm now loading the new handles. 

Going to continue on this tomorrow.

BastiaanOlij commented:
OK, this is starting to work. All the basic buttons and haptics are working on the left and right controllers.

Because the action system is declarative no buttons will work on any additional controllers.
I am planning on doing something more with that, we'll see.

Last thing I'll be working on before this gets merged is custom actions. 
So you will be able to add your own actions to the manifest file and be alerted of those.

BastiaanOlij commented:
I've added actions for analog feedback of the triggers, added a proper default bindings file for the oculus touch (I will be doing index later today) and improved the demo a little bit so it shows feedback of one of the controllers (might change the project to show them for both).

BastiaanOlij commented:
I have to admit, I'm impressed with the interface Valve came up with here. If you make a mistake in your configuration files it does fail without really letting your know whats up, but once it's up and running it rocks. I managed to configure our default actions for all supported controllers. 

Obviously the Vive wands and Windows controllers don't have A and B buttons so they won't action anything.

![image](https://user-images.githubusercontent.com/1945449/76159698-31a32780-6177-11ea-9336-5654bc4988b5.png)

I also restructured a few things in the demo project so we have more scripts to get people up and running quickly.

Finally rendering is now done to a dedicated viewport while I'm outputting the preview with a texture rect to screen. This texture rect applies a linear to sRGB conversion so the colors are correct both in the headset and on screen.

BastiaanOlij commented:
Ok, I've added two new GDNative classes.

OpenVRPose is a spatial node that you can bind to any pose you've defined in your actions.json
So say you add:
```
    {
      "name": "/actions/godot/in/left_hand_grip",
      "type": "pose"
    },
```
And you bind that to the left hand grip pose slot than you can use `/actions/godot/in/left_hand_grip` as the action property in the inspector:
![image](https://user-images.githubusercontent.com/1945449/76162254-109b0080-6190-11ea-9133-acd94d5f23f4.png)

When you do this the position of the spatial node will be automatically updated. Just like the other ARVR nodes that are being positioned this node must be a child of ARVROrigin!

The other new class is OpenVRAction which has two bind slots. The first is a pressed action, the second an analog action. So you can add the following to the actions.json:
```
    {
      "name": "/actions/godot/in/custom_boolean",
      "type": "boolean"
    },
    {
      "name": "/actions/godot/in/custom_analog",
      "type": "vector2"
    }
```
And then bind these to buttons/joysticks on your controller. 
![image](https://user-images.githubusercontent.com/1945449/76162298-9454ed00-6190-11ea-9c56-28de7355dbc6.png)

You can place the OpenVRAction node anywhere you like in the scene. The button presses will emit signals on this node but the joystick position needs to be read with get_analog which returns a Vector2 (you can bind this to say your trigger and use `"type": "vector1"` in actions.json and only the x of the vector will be populated).

I'll be added two more classes soon:
OpenVRHaptic which gives you complete control over a haptic output
OpenVRSkeleton which will return a hand skeleton based on the skeletal logic in OpenVR.

beniwtv commented on src/ARVRInterface.cpp:
```diff
@@ -226,12 +158,15 @@ godot_transform godot_arvr_get_transform_for_eye(void *p_data, godot_int p_eye,
 	ret = *p_cam_transform;
 	ret = godot::api->godot_transform_operator_multiply(&ret, &reference_frame);
-	ret = godot::api->godot_transform_operator_multiply(&ret, &arvr_data->hmd_transform);
+	ret = godot::api->godot_transform_operator_multiply(&ret, &arvr_data->ovr->get_hmd_transform());
```
This fails compiling for me with error:

```
src/ARVRInterface.cpp: In function 'godot_transform godot_arvr_get_transform_for_eye(void*, godot_int, godot_transform*)':
src/ARVRInterface.cpp:161:95: error: taking address of temporary [-fpermissive]
  ret = godot::api->godot_transform_operator_multiply(&ret, &arvr_data->ovr->get_hmd_transform());
```
I've solved it by assigning arvr_data->ovr->get_hmd_transform() to a variable first.

BastiaanOlij replied:
Sounds like a similar issue like we had with returning vectors....

BastiaanOlij replied:
I'm returning it as a const pointer to the variable, lets see if that works.

beniwtv replied:
Yep seems to work.

beniwtv commented on src/ARVRInterface.cpp:
```diff
@@ -330,164 +268,39 @@ void godot_arvr_commit_for_eye(void *p_data, godot_int p_eye, godot_rid *p_rende
 	arvr_data->ovr = openvr_data::retain_singleton();
 	arvr_data->video_driver = 0;
-	godot::api->godot_transform_new_identity(&arvr_data->hmd_transform);
+	godot::api->godot_transform_new_identity(&arvr_data->ovr->get_hmd_transform());
```
Same error about temporaries here.

BastiaanOlij replied:
This line can actually be removed, it's already properly initialized inside of the openvr_data object.

beniwtv commented on src/open_vr/openvr_data.cpp:
```diff
@@ -264,6 +395,406 @@ void openvr_data::get_eye_to_head_transform(godot_transform *p_transform, int p_
+// Bind our default actions, if the actions file has been edit
+// and the defaults removed they will simply be undefined
+void openvr_data::bind_default_action_handles(tracked_device *p_device, godot_int p_hand) {
+	char *hand = p_hand == 1 ? "left" : "right";
```
Got the following compile error here:

```
src/open_vr/openvr_data.cpp: In member function 'void openvr_data::bind_default_action_handles(openvr_data::tracked_device*, godot_int)':
src/open_vr/openvr_data.cpp:561:27: error: invalid conversion from 'const char*' to 'char*' [-fpermissive]
  char *hand = p_hand == 1 ? "left" : "right";
               ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
```
I solved that by writing const char *hand instead.

beniwtv commented on src/open_vr/openvr_data.cpp:
```diff
@@ -264,6 +395,406 @@ void openvr_data::get_eye_to_head_transform(godot_transform *p_transform, int p_
+		"button_ax",
+		"button_by"
+	};
+	char *output_actions[DAH_OUT_MAX] = {
```
This gives a warning (not an error):

"warning: ISO C++ forbids converting a string constant to 'char*'"

BastiaanOlij replied:
Adding const to all of these

BastiaanOlij commented:
@beniwtv this latest commit should fix your problems though I am not 100% sure if the hmd_transform one works. That one is weird.

BastiaanOlij commented:
I've added support for a new haptics class. Just add a spatial node somewhere in your scene and add the OpenVRHaptic.gdns script to it. Now you can set the output action which should be defined like so:
```
    {
      "name": "/actions/godot/out/left_haptic",
      "type": "vibration"
    },
```
Set the duration, frequency and amplitude, and call trigger_pulse on the object.
I could have made this an Object and just purely do it in script but I like the visual interaction. 

I've also added some logic to the demo where the controllers now have a collision shape that follow the controllers around BUT they don't go through solid objects! When a collision is detected it moves no further and I use the new haptic function to trigger a pulse. You can see it in action here:
https://youtu.be/oRc1pTUI_CQ
https://youtu.be/PIdCYWW-R_0

The blue sphere is purely there to visualize what is happening.

BastiaanOlij commented:
Tomorrow I start working on documentation, I'm looking to merge this pretty much as we have it right now unless I get feedback people don't like it.
I'm doing the hand skeleton work in a separate PR

ChristophHaag commented:
> If you make a mistake in your configuration files it does fail without really letting your know whats up

If it's a json parse error, it should show up in `steam/logs/vrserver.txt`, or wherever that log file is on windows.

> Obviously the Vive wands and Windows controllers don't have A and B buttons so they won't action anything.

That's where the problem with the "everything has to look like an xbox 360 controller" style input comes in.

The intention for the action system in OpenVR and OpenXR was really to abandon hardcoding a default controller layout and replace it with directly exposing actions to game developers.

For a game engine like Godot that would mean a UI that exposes 1. the ability to create an action ("Teleport") with a type ("Boolean") and 2. a bunch of known controllers with their available inputs ("Index Controller: trigger/click", "Oculus Rift Controller: trigger/click").

The action and binding json manifests for SteamVR could then be dynamically generated when building the project.

It doesn't look like a huge advantage at first glance, but a lot of people are huge fans of this approach.

For example if you try the "Show bindings on controllers in VR" option from SteamVR, you are supposed to see **what** each of the buttons does via action name/description as you touch them:
![image](https://user-images.githubusercontent.com/781076/76249093-e6eee180-6242-11ea-89b0-d380d55b0ee4.png)

> Because the action system is declarative no buttons will work on any additional controllers.

This responsibility is pushed into SteamVR or OpenXR runtimes. In SteamVR users can see all the actions a game uses, and can bind whatever inputs they have to the actions. Those bindings can then be shared with others in the Steam workshop.

beniwtv commented:
Everything including overlay works for me with the latest code

BastiaanOlij commented:
Hey @ChristophHaag , 

> The intention for the action system in OpenVR and OpenXR was really to abandon hardcoding a default controller layout and replace it with directly exposing actions to game developers.

Yup, the problem I'm dealing with is building an engine that supports OpenVR, OpenXR and many others, many of which don't have an action system. So making it possible to deploy to multiple platforms with suddenly needing vastly different approaches is a pain. 

The action system is growing on me though, especially considering people only have two hands it really makes sense.

Right now the default actions that are in the action.json file basically allow you to mimic the old approach having actions for each old button. But it's meant as a stepping stone or for people who indeed want to deploy to a Quest and the Oculus Store and to SteamVR and can't use an action system that only works on one system.

If you DO want to use the action system as intended, trash the action system and start using the new functionality. For now as OpenVR only objects but once I sink my teeth into OpenXR (Still have to start playing with your plugin) I'll start looking at how we can feed some of this logic back into the core for Godot 4 (3.2 is feature locked now so). 

The only thing I'm not happy about currently is the way OpenVRAction works. I feel this needs to be integrated into a ARVRController subclass. But OpenVRPose and OpenVRHaptic I think are the right approach.

BastiaanOlij commented:
@ChristophHaag btw, one thing I can't wrap my head around is how a "fire gun" action would work. You bind it to your left hand trigger and you pick your gun up with your right hand...

rpavlik commented:
So this is where sub-action paths come in. I'd have to double-check my notes to explain exactly how, but this is the exact reason for sub-action paths.

Wallbraker commented:
> @ChristophHaag btw, one thing I can't wrap my head around is how a "fire gun" action would work. You bind it to your left hand trigger and you pick your gun up with your right hand...

Hello I'm Jakob and work on Moando a OpenXR runtime (and is a member of the Khronos OpenXR WG, but only speak for myself). So I can not speak for it is done in OpenVR as I have no experience with it. But in OpenXR you would use sub action groups, where you can ask for the state of action with either sub-action `/user/hand/left`, `/user/hand/right` or none.

rpavlik commented:
thank you Jakob - It's been a while since I reviewed the subaction stuff so I didn't want to say too much more to avoid getting it wrong :D  (I'm also a Monado dev, and the OpenXR spec editor but as with Jakob, I'm not speaking for the group as a whole most of the time)

ChristophHaag commented:
In xrdesktop at first we had duplicate actions for each hand, but then SteamVR did introduce a way to assign one action to both hands, and poll action data while restricting to certain device paths

First in the actions.json you may need something like
```json
   "action_sets": [
      {
         "name": "/actions/godot",
         "usage": "leftright"
      }
    ]
```
with `"usage": "leftright"`, and the same action bound to both left and right hands.

And then the actions would be polled once per hand or device. If you want to have it completely dynamic, you can query those first (somewhat pseudocode):
```c++
  VRInputValueHandle_t *origin_handles = malloc(sizeof(VRInputValueHandle_t) * k_unMaxActionOriginCount);
  EVRInputError err = VRInput()->GetActionOrigins(actionset_handle, action_handle, origin_handles, k_unMaxActionOriginCount);
  
  int origin_count = -1;
  while (origin_handles[++origin_count] != k_ulInvalidInputValueHandle);
  
  for (int i = 0; i < origin_count; i++) {
    InputOriginInfo_t origin_info;
    err = VRInput()->GetOriginTrackedDeviceInfo(origin_handles[i], &origin_info, sizeof (origin_info));
    
    devicepath_list.append(origin_info.devicePath);
  }
```

And then when polling
```c++
  for (VRActionHandle_t devicepath: devicepath_list) {
    InputAnalogActionData_t data;
    VRInput()->GetAnalogActionData(action_handle, &data, sizeof(data), devicepath);
    
    // also possible:
    InputOriginInfo_t origin_info;
    VRInput()->GetOriginTrackedDeviceInfo(data.activeOrigin, &origin_info, sizeof (origin_info));
  }
```

BastiaanOlij commented:
Thanks for the feedback guys. I'm going to write up some documentation of what I've build so far tonight but I think I'm going to change the objects to allow filtering on hand and probably look into doing OpenVRActions differently and just combine that logic into an ARVRController subclass.

BastiaanOlij commented:
Ok, simplified a bunch of things and made it possible to bind OpenVRAction, OpenVRPose and OpenVRHaptic to a either hand or have them work on any.

Also removed the left/right hand duplication of actions in the default actions.json file. You can bind the same default action to both controllers and Godot will be able to ensure actions react to the right controller. 

It does seem that OpenVR does not have any hints in the actions.json file to indicate to the user if the action has to be bound to a specific controller or can be used on any controller.

BastiaanOlij commented:
I've made a start on documenting the new system:

https://github.com/GodotVR/godot-openvr-asset/wiki/OpenVR-actions

BastiaanOlij commented:
I'm happy with how this turned out so far, documentation is pretty much up to date now as well:
https://github.com/GodotVR/godot-openvr-asset/wiki/OpenVR-actions

I'm planning to merge this sometime this week so I can start working on finger tracking.

Any feedback is very welcome.

MathiasBaumgartinger commented:
I honestly have no idea on the complications that this would cause, but do you see any possibility to include the logic of the action-files in the native input map of godot?
